packages/itwinui-react/src/core/TransferList/TransferList.tsx

```diff
@@ -0,0 +1,276 @@
+    >
+      {children}
+    </Element>
+  );
+}) as PolymorphicForwardRefComponent<'div', TransferListAreaOwnProps>;
+
+// ----------------------------------------------------------------------------
+// TransferList.List component
+
+type TransferListListOwnProps = ListProps; // eslint-disable-line @typescript-eslint/ban-types
+
+export type TransferListListProps<T extends React.ElementType = 'div'> =
+  PolymorphicComponentProps<T, TransferListListOwnProps>;
+
+const TransferListList = React.forwardRef((props, ref) => {
+  const { as: Element = 'ul', children, ...rest } = props;
+
+  const [focusedIndex, setFocusedIndex] = React.useState<number | null>();
+  const listRef = React.useRef<HTMLUListElement>(null);
+  const refs = useMergedRefs(listRef, ref);
+
+  const getFocusableNodes = React.useCallback(() => {
+    const focusableItems = getFocusableElements(listRef.current);
+    // Filter out focusable elements that are inside each list item, e.g. checkbox, anchor
+    return focusableItems.filter(
+      (i) => !focusableItems.some((p) => p.contains(i.parentElement)),
+    ) as HTMLElement[];
+  }, []);
+
+  React.useEffect(() => {
+    const items = getFocusableNodes();
+    if (focusedIndex != null) {
+      (items?.[focusedIndex] as HTMLLIElement)?.focus();
+      return;
+    }
+  }, [focusedIndex, getFocusableNodes]);
+
+  const onKeyDown = (event: React.KeyboardEvent<HTMLUListElement>) => {
+    if (event.altKey) {
+      return;
+    }
+
+    const items = getFocusableNodes();
+    if (!items?.length) {
+      return;
+    }
+
+    const currentIndex = focusedIndex ?? 0;
+
+    switch (event.key) {
+      case 'ArrowDown': {
+        setFocusedIndex(Math.min(currentIndex + 1, items.length - 1));
+        event.preventDefault();
+        event.stopPropagation();
+        break;
+      }
+      case 'ArrowUp': {
+        setFocusedIndex(Math.max(currentIndex - 1, 0));
+        event.preventDefault();
+        event.stopPropagation();
+        break;
+      }
+      default:
+        break;
+    }
+  };
+
+  return (
+    <List
+      as={Element}
+      onKeyDown={onKeyDown}
+      role='listbox'
+      ref={refs}
+      {...rest}
+    >
+      {children}
+    </List>
+  );
+}) as PolymorphicForwardRefComponent<'ul', TransferListListOwnProps>;
+
+// ----------------------------------------------------------------------------
+// TransferList.ListItem component
+
+type TransferListListItemOwnProps = {
+  /**
+   * Callback fired when the the active state changes.
+   */
+  onActiveChange?: (value: boolean) => void;
+} & ListItemProps;
+
+export type TransferListListItemProps<T extends React.ElementType = 'li'> =
+  PolymorphicComponentProps<T, TransferListListItemOwnProps>;
+
+const TransferListListItem = React.forwardRef((props, ref) => {
+  const { actionable, disabled, onActiveChange, children, active, ...rest } =
+    props;
+
+  const onClickEvents = () =>
+    actionable && onActiveChange && onActiveChange(active ?? false);
```

Comment:
why is `onActiveChange` called with the stale state? I would expect it to be called with the next expected state